pkg/controller/shared_state.go

```diff
@@ -1564,3 +1571,60 @@ func (c *SharedState) getNodesTopologySelector(ctx context.Context, nodes []v1.N
 	}
 	return topologySelector
 }
+
+func (s *SharedState) addNodeToAvailableAttachmentsMap(ctx context.Context, node string) {
+	nodeObj := &v1.Node{}
+	if err := s.cachedClient.Get(ctx, types.NamespacedName{Name: node}, nodeObj); err != nil {
+		capacity, _ := azureutils.GetNodeMaxDiskCountWithLabels(nodeObj.Labels)
+		var count atomic.Int32
+		count.Store(int32(capacity))
+		s.availableAttachmentsMap.Store(node, &count)
+		return
+	}
+	capacity, _ := azureutils.GetNodeRemainingDiskCount(ctx, s.cachedClient, node)
+	var count atomic.Int32
+	count.Store(int32(capacity))
+	s.availableAttachmentsMap.Store(node, &count)
+}
+
+func (s *SharedState) deleteNodeFromAvailableAttachmentsMap(ctx context.Context, node string) {
+	w, _ := workflow.GetWorkflowFromContext(ctx)
+	w.Logger().V(5).Infof("Delete node (%s) from availableAttachmentsMap", node)
+	s.availableAttachmentsMap.Delete(node)
+}
+func (s *SharedState) decrementAttachmentCount(ctx context.Context, node string) bool {
+	w, _ := workflow.GetWorkflowFromContext(ctx)
+	remainingCapacity, nodeExists := s.availableAttachmentsMap.Load(node)
+	if nodeExists && remainingCapacity != nil {
+		currentCapacity := int32(0)
+		for {
+			currentCapacity = remainingCapacity.(*atomic.Int32).Load()
+			if currentCapacity == int32(0) || remainingCapacity.(*atomic.Int32).CompareAndSwap(currentCapacity, currentCapacity-1) {
+				if currentCapacity == int32(0) {
+					err := status.Error(codes.OutOfRange, fmt.Sprintf("Current Capacity of disk attachment is 0. Can't decrement attachment count anymore."))
+					w.Logger().Errorf(err, "failed to decrement attachment count for node (%s) ", node)
```

Comment:
Rather the log a formatted string, use `Info` and pass the node as a key value pair, e.g.

```go
w.Logger().Error(err, "Failed to decrement attachment count", consts.NodeNameLabel, node)
```